akka-actor-typed/src/main/scala/akka/actor/typed/internal/routing/PoolRouterImpl.scala

```diff
@@ -30,6 +30,12 @@ private[akka] final case class PoolRouterBuilder[T](
 
   def withRoundRobinRouting(): PoolRouterBuilder[T] = copy(logicFactory = () => new RoutingLogics.RoundRobinLogic[T])
 
+  def withConsistentHashingRouting(
+    virtualNodesFactor: Int = 0,
```

Comment:
In a pool the set of children will only change if a child stops, and that is a bad scenario, there is no adding children going on, so maybe always use 1 for pools and only have it configurable for groups where actors can come and go instead.

Reply:
TBH, I do not understand connection between fact that pool uses, well, pool of actors and `virtualNodesFactor`. 

To best of my understanding `virtualNodesFactor` is used so that hash will spread more evenly. Let's say we have `n` actors in pool. Then with virtual nodes factor equal to 1, we will have `n` points on the hashing ring - each one of them corresponding to an actor. But if we increase virtual nodes factor then every actor from will be mapped to `n` points - making sure that load is spread more evenly.

Am I missing something in this case?

Reply:
With a fixed set of routees, like in a pool, and a good hashing algo (with good distribution) you will get good spread of the load. 

My understanding of idea of virtual nodes being useful is that when you add and remove routees, like what can happen in a group router, to get more uniform re-distribution across the hash wheel. See the *Demonstration* part of the article you referenced: http://www.tom-e-white.com/2007/11/consistent-hashing.html

Reply:
I've read this article before and now re-read again and I get different understanding, please correct me if I am wrong.

> With a fixed set of routees, like in a pool, and a good hashing algo (with good distribution) you will get good spread of the load.

That is quite an assumption :).

Overall routees hashing happens with help of the `akka.scala.routing.ConsistentHash`:
```
  def apply[T: ClassTag](nodes: Iterable[T], virtualNodesFactor: Int): ConsistentHash[T] = {
    new ConsistentHash(
      immutable.SortedMap.empty[Int, T] ++
      (for {
        node <- nodes
        nodeHash = hashFor(node.toString)
        vnode <- 1 to virtualNodesFactor
      } yield (concatenateNodeHash(nodeHash, vnode) -> node)),
      virtualNodesFactor)
  }
```
`hashFor` uses `scala.akka.routing.MurmurHash`, which I think we can consider good.

From my understanding virtual nodes are useful if you have low amount of actor ref in pool, even in the aformentioned article, the relatively good and equal spread across hash ring is achieved starting at around _100_ replicas.

Maybe I'll try to reiterate my understanding, so it's clearer.

1. We have a hash ring.
2. Our ActorRefs will be put as points around the hash ring using `MurmurHash`.
3. We do not have full control over `MurmurHash` nor the `toString` representation of `ActorRef`s
4. Virtual Nodes will still point to the original set of actor refs
5. Nodes (virtual and actual) will gather request starting from the closest node (going counter-clockwise)
6. The bigger the amount (to a degree of course) the more even the load will be spread.

All of the above leads me to believe that we should leave possibility of setting `virtualNodesFactor` in pool.

Reply:
Alright, you convinced me, let's keep the virtual nodes for pool as well!